### ISSUE TITLE
VB: commenting then un-commenting a block of code yields a different end-result

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#14308]
If I comment, then un-comment a block like below, the end result is different than what I started with.

I’m wondering whether that is fixable though. C# allows for comments at the start of the line, but in VB they automatically get re-aligned…

Before:

``` VB
            Dim source1 =
<compilation>
    <file name="c.vb"><![CDATA[
Public Class C
    Implements I

    Public Property P As (a As Object, b As Object) Implements I.P
End Class
]]></file>
</compilation>
```

After:

``` VB

            Dim source1 =
<compilation>
    <file name="c.vb"><![CDATA[
            Public Class C
                Implements I

                Public Property P As (a As Object, b As Object) Implements I.P
            End Class
            ]]></file>
</compilation>
```

FYI @CyrusNajmabadi

**Comment by CyrusNajmabadi** [comment]
can you show what you get if you do the comment/uncomment?

**Comment by jcouv** [comment]
Updated the issue to show the incorrect end-result.

**Comment by CyrusNajmabadi** [comment]
IIRC, CData is whitespace sensitive.  So this is definitely bad behavior.

**Comment by Pilchie** [comment]
All XML is potentially whitespace sensitive, though I believe the compiler has code to elide shared indentation (@AnthonyDGreen, is that right)?

Still, when we uncomment, we're adjusting the indentation of the first line with the `Dim`, so we should be preserving the relative indent of the rest of the lines too.

**Comment by Pilchie** [comment]
Note - this also affects string literals in VB now that they can be multi-line.

**Comment by AnthonyDGreen** [comment]
@Pilchie I don't believe we elide the shared indentation, as glorious as that would be.

**Comment by CyrusNajmabadi** [comment]
I need @heejaechang 's help with this.  The issue appears to be as follows.

First, we comment out the selected text.  The comment service transforms

```vb
    Sub M()
            Dim source1 =
<compilation>
    <file name="c.vb"><![CDATA[
Public Class C
    Implements I

    Public Property P As (a As Object, b As Object) Implements I.P
End Class
]]></file>
</compilation>
End Sub
```

Into:

```vb
    Sub M()
'        Dim source1 =
'<compilation>
'    <file name="c.vb"><![CDATA[
'Public Class C
'    Implements I

'    Public Property P As (a As Object, b As Object) Implements I.P
'End Class
'        ]]></file>
'</compilation>
End Sub
```

Note that all the code is left aligned.  After this happens, code cleanup takes over and formats this code.  That transforms the above into:

```vb
    Sub M()
        '        Dim source1 =
        '<compilation>
        '    <file name="c.vb"><![CDATA[
        'Public Class C
        '    Implements I

        '    Public Property P As (a As Object, b As Object) Implements I.P
        'End Class
        '        ]]></file>
        '</compilation>
    End Sub
```

Note: this indentation is important.  If we don't indent, and then we uncomment, then we get the right result.

Now, when we uncomment the above code, we then have a tree like so:

```vb
    Sub M()
                Dim source1 =
        <compilation>
            <file name="c.vb"><![CDATA[
        Public Class C
            Implements I

            Public Property P As (a As Object, b As Object) Implements I.P
        End Class
                ]]></file>
        </compilation>
    End Sub
```

This then gets formatted.  The formatted pulls 'Dim Source1' over to the left, and carries the sub-expression along with it.  *Except* that it doesn't pull along the actual text in the xml expression, only the xml elements.  That leads us with:


```vb
    Sub M()
        Dim source1 =
<compilation>
    <file name="c.vb"><![CDATA[
        Public Class C
            Implements I

            Public Property P As (a As Object, b As Object) Implements I.P
        End Class
                ]]></file>
</compilation>
    End Sub
```

Voila.  Comment/Uncomment leads to the code inside the xml element being indented.

@heejaechang There are a lot of moving parts inside the formatter that are happening here.  It seems to me that we probably want the inner text to be anchored to the containing element (maybe)? I'm not familiar enough with all the formatter concepts to know what the right thing to do here is.

Could you explain what you think is an appropriate fix here, and where you think it should go?

Thanks!

**Comment by heejaechang** [comment]
I believe vb commit formatter rule is written to never touch anything inside of xml literal (CDATA)

I think same issue will happen to csharp verbatim string. 

basically, since formatter can't know the original state of the code before it was commented. when it is commented, formatter see code as it is, and leave literal text as they are.

**Comment by CyrusNajmabadi** [comment]
I think a strange thing is that VB auto-indents the commented out code.  C# does not do that.  Is there a way we can get VB to stop doing that?

**Comment by heejaechang** [comment]
2 possibility. vb code comment feature call formatting explicitly or vb commit formatter recognize commenting and moving caret as a cue to do auto formatting.

C# doesn't have any auto formatting except ";" or "}" that probably why it doesn't auto format.

is pre-roslyn era vb auto formatter doesn't do what it is doing now?

**Comment by CyrusNajmabadi** [comment]
Interestingly enough, this reproes for C# *if* you format between the comment and uncomment steps.  You can see this if you attempt to comment the string literal in this code:

```c#
public class Foo
{
    public Foo()
    {
        var v = @"
foo
bar
baz
quux
";
    }
}
```

I'm *leaning* toward thinking that the bug is that we indent *left aligned* comments when formatting.  If we didn't do that, tehn we wouldn't have this problem.

Alternatively, we can keep things are they are today, *except* we can attempt to make it so that commit-cleanup doesn't run when commenting is done.  This would help the situation out as long as the user did not then format the document.

Will have to noodle more on htis.

**Comment by heejaechang** [comment]
the reason formatter doesn't change anything inside of the literal is since it is actually changing result of the code. formatter should never change meaning of the code it is given.

when user did comment code feature, they changed the meaning of code and then when it uncomment the code, it introduces new code to code base which is different than original code. formatter doesn't know the original code, it only see the new code introduced and it will format them following rules which doesn't change meaning of the given input. 

moving content of the string literal changes meaning of the code.

I will say this is won't fix. but if we want to pursue fixing this, I think focus should be on the comment/uncomment feature not formatting.

one way of comment/uncomment feature can fix this is, comment feature remember original code, and then uncomment happen, put the original code back.

**Comment by CyrusNajmabadi** [comment]
I agree this is not formatting's fault *when it uncomments*.  However, i think that formatting indents a commented out section is problematic, (though i would not use the term 'fault').  THe issue is that we tend to think of 'comments' as being effectively meaningless.  As such, it's safe to move them.  As we've discovered here, when the comments are actually commented-out code, and the user wants to restore the original code, our decision has negatively affected their ability to do that.

Here are some ideas (in no order of preference):
1.  When commenting, look at the orignal document range that was commented.  If it intersects a multi-line, whitespace sensitive construct (i.e. xml literals, multiline strings), then do whatever is necessary so that the formatter doesn't run and indent that comment.  That will ensure that if the user then uncomments, they can get back to their original code.
2. The formatter runs on un-comment because the commenting engines calls it explicitly.  Given that, the commenting engine can say "you are formatting a range of code that was uncommented.  as such, you are permitted to touch the whitespace in a literal".
3. If comments are left-aligned, do not have the formatter touch them.

**Comment by heejaechang** [comment]
it is up to IDE, but figuring out user's intention can be really hard. it might satisfy one user case but might upset other users like I just commented out this code and format why it doesn't go to right place? or I just uncommented this code that I didn't comment or I hand commented, they it is messing with my string literal?

if problem starts from commenting feature to uncommenting feature, I think state tracking should start with commenting feature invocation.
